### PR TITLE
RayCluster integration with Kueue 

### LINKED ISSUE #1272: Add support for RayCluster
**What would you like to be added**:

Support RayCluster as a queue-able workload in Kueue (much like RayJob).

**Why is this needed**:

Currently Kueue supports RayJob which works great when managing ray jobs that run on ephemeral ray clusters. However, there are many use-cases and existing workloads that depend on long-lived RayClusters. Being able to account for these RayClusters with Kueue would greatly improve integration of Kueue with Ray.

**Completion requirements**:

This probably needs a KEP, but very roughly the requirements would be:
* Kueue should only account for ray worker nodes in ClusterQueue quotas (we may account for head nodes later)
* Scaling of worker nodes should be blocked on quota reservations
* Evicting or pre-empting RayClusters involves scaling worker groups to 0

This enhancement requires the following artifacts:

- [X] Design doc
- [X] API change
- [X] Docs update

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by vicentefb**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Support RayCluster as a queue-able workload in Kueue since there are many use-cases and existing workloads that depend on long-lived RayClusters.

#### Which issue(s) this PR fixes:
Fixes #1272

#### Special notes for your reviewer:
This implementation is using master's version of kuberay and using `v1` api version instead of `v1alpha1` because it's using the new suspend API (https://github.com/ray-project/kuberay/issues/1667) for RayCluster

#### Does this PR introduce a user-facing change?
```release-note
Support RayCluster as a queue-able workload in Kueue
```

### CODE REVIEW

**Comment by vicentefb** [comment]
Will squash commits once final review is given.

File: pkg/controller/jobs/raycluster/raycluster_webhook.go
Comment (andrewsykim) [inline pkg/controller/jobs/raycluster/raycluster_webhook.go:90]: Add a TODO here to revisit this once https://github.com/kubernetes-sigs/kueue/issues/77 is implemented

File: test/integration/controller/jobs/raycluster/raycluster_controller_test.go
Comment (andrewsykim) [inline test/integration/controller/jobs/raycluster/raycluster_controller_test.go:121]: return true?

File: test/integration/controller/jobs/raycluster/raycluster_controller_test.go
Comment (andrewsykim) [inline test/integration/controller/jobs/raycluster/raycluster_controller_test.go:122]: BeTrue()?

File: test/integration/controller/jobs/raycluster/raycluster_controller_test.go
Comment (andrewsykim) [inline test/integration/controller/jobs/raycluster/raycluster_controller_test.go:147]: You can uncomment this test case and update go modules for this branch to point to master. The PR will be blocked anyways until kuberay v1.1.0 is released anyways

**Comment by andrewsykim** [comment]
/hold 

Holding until there's a kuberay release that has https://github.com/ray-project/kuberay/pull/1711

File: pkg/controller/jobs/raycluster/raycluster_controller_test.go
Comment (alculquicondor) [inline pkg/controller/jobs/raycluster/raycluster_controller_test.go:38]: Let's just have `TestReconcile`. These other tests were added for other integration before we found a way to test at the reconcile level.
Make sure to move the test cases.

File: pkg/util/testingjobs/raycluster/wrappers.go
Comment (alculquicondor) [inline pkg/util/testingjobs/raycluster/wrappers.go:33]: maybe `MakeCluster` and `ClusterWrapper` would be less confusing.

File: pkg/controller/jobs/raycluster/raycluster_controller_test.go
Comment (alculquicondor) [inline pkg/controller/jobs/raycluster/raycluster_controller_test.go:291]:
```suggestion
		"when workload is admitted, cluster is unsuspended": {
```

File: pkg/controller/jobs/raycluster/raycluster_controller_test.go
Comment (alculquicondor) [inline pkg/controller/jobs/raycluster/raycluster_controller_test.go:463]: Uncomment here an below

File: pkg/controller/jobs/raycluster/raycluster_controller.go
Comment (astefanutti) [inline pkg/controller/jobs/raycluster/raycluster_controller.go:52]: Some permissions on the `finalizers` sub-resource should be added too (see #1485), e.g.:

```
// +kubebuilder:rbac:groups=ray.io,resources=rayclusters/finalizers,verbs=get;update
```

File: pkg/controller/jobs/raycluster/raycluster_controller.go
Comment (astefanutti) [inline pkg/controller/jobs/raycluster/raycluster_controller.go:46]: Should `IsManagingObjectsOwner` be provided as well?

At the moment there is a "hack" in the Pod controller, inherited from the RayJob support:

https://github.com/kubernetes-sigs/kueue/blob/1579d75e5a35bdb4ab24c64b2da4e4effa98bbc8/pkg/controller/jobs/pod/pod_controller.go#L817-L822

That won't work for v1, and it may be better to take that opportunity to have that logic provided by the RayCluster integration itself.

**Comment by andrewsykim** [comment]
@alculquicondor @vicentefb can we try to land this for v0.6.0?

**Comment by vicentefb** [comment]
Creation of a RayJob to show that there's no double-queueing (i.e. no workload created for its ray cluster):
image

Logs from Kueue:
image

Creation of a RayCluster resource:
image

Logs from Kueue:
image

File: pkg/controller/jobframework/reconciler.go
Comment (astefanutti) [inline pkg/controller/jobframework/reconciler.go:247]: I wonder if calling `IsOwnerManagedByKueue` here would generalise it. Otherwise, I'd suggest to add a TODO so we can work on generalising that logic. I think this would also apply to #1568.

File: go.mod
Comment (andrewsykim) [inline go.mod:16]: FYI @alculquicondor @vicentefb this is still pointing to master. Last time I spoke with @kevin85421 the next release won't be until ~Feb, are we okay merging latest version of the module for now and updating this to v1.1 or should we block on it?

**Comment by astefanutti** [comment]
I've tested it locally for few happy paths and it works like a charm. Great job @vicentefb!

File: go.mod
Comment (alculquicondor) [reply go.mod:16]: I'm ok moving forward like this, but we can probably fix the dependency on a path release, once ray-operator is released.

**Comment by alculquicondor** [comment]
Please fix verify and address the comments from @astefanutti

File: pkg/controller/jobframework/reconciler.go
Comment (astefanutti) [inline pkg/controller/jobframework/reconciler.go:246]: It seems those two comment lines would better be kept close to `r.ensureOneWorkload(ctx, job, object)`.

File: pkg/controller/jobs/raycluster/raycluster_controller.go
Comment (astefanutti) [inline pkg/controller/jobs/raycluster/raycluster_controller.go:191]: Is `|| jobframework.IsOwnerManagedByKueue(owner)` actually needed here?

**Review by astefanutti — APPROVED** [review]
Left two non-essential comments. Otherwise LGTM 👍🏼. Thanks!

File: client-go/clientset/versioned/scheme/register.go
Comment (alculquicondor) [inline client-go/clientset/versioned/scheme/register.go:29]: This shouldn't be necessary. Please revert.

File: pkg/controller/jobframework/reconciler.go
Comment (alculquicondor) [inline pkg/controller/jobframework/reconciler.go:38]: remove this empty line

File: pkg/controller/jobframework/reconciler.go
Comment (alculquicondor) [inline pkg/controller/jobframework/reconciler.go:251]: Let's try to make the implementation the same as for Job/JobSet.

In job_webhook.go, we add a parent workload annotation to the Job object.

We can do the same: add a parent workload annotation to the RayCluster if it's owned by a RayJob.

File: pkg/controller/jobs/raycluster/raycluster_controller.go
Comment (alculquicondor) [reply pkg/controller/jobs/raycluster/raycluster_controller.go:191]: Actually, it could be the other way, we just need to check `jobframework.IsOwnerManagedByKueue(owner)`.
Then in the future we don't need to change this line if there is another CRD managed by Kueue that can create RayClusters.

File: test/integration/controller/jobs/raycluster/raycluster_controller_test.go
Comment (alculquicondor) [inline test/integration/controller/jobs/raycluster/raycluster_controller_test.go:70]:
```suggestion
var _ = ginkgo.Describe("RayCluster controller", ginkgo.Ordered, ginkgo.ContinueOnFailure, func() {
```

File: test/integration/framework/framework.go
Comment (alculquicondor) [inline test/integration/framework/framework.go:30]: Remove this. We don't need to use v1 yet.

Comment (alculquicondor) [reply test/integration/framework/framework.go:30]: Nvm... It seems that using v1alpha1.RayJob and v1.RayCluster is working fine.

File: pkg/controller/jobs/raycluster/raycluster_controller.go
Comment (astefanutti) [reply pkg/controller/jobs/raycluster/raycluster_controller.go:191]: Right. Could that logic be generalised and applicable to any integration then?

Comment (alculquicondor) [reply pkg/controller/jobs/raycluster/raycluster_controller.go:191]: It could. We can make it part it the job framework webhook helpers. But I'm fine leaving such thing for a follow up. Just matching the implementation for now should be enough.

File: test/integration/controller/jobs/raycluster/raycluster_controller_test.go
Comment (astefanutti) [inline test/integration/controller/jobs/raycluster/raycluster_controller_test.go:161]: This is not the same scheme as the test scheme where types are registered into.

The `sigs.k8s.io/kueue/client-go/clientset/versioned/scheme` import should be replaced with `k8s.io/client-go/kubernetes/scheme` instead.

File: test/integration/controller/jobs/raycluster/raycluster_controller_test.go
Comment (alculquicondor) [inline test/integration/controller/jobs/raycluster/raycluster_controller_test.go:568]: Please add a second workload that initially doesn't fit, until the first RayCluster is terminated (deleted?) by the user (the user is the test code, in this case).

**Comment by vicentefb** [comment]
To track kuberay dependency release version when available https://github.com/kubernetes-sigs/kueue/issues/1634

File: test/integration/controller/jobs/raycluster/raycluster_controller_test.go
Comment (alculquicondor) [inline test/integration/controller/jobs/raycluster/raycluster_controller_test.go:598]: Users should not have to delete the workload. This is usually done by the garbage collector, but there is no garbage collection in integration tests. Please add a comment for that.

File: pkg/controller/jobs/raycluster/raycluster_controller_test.go
Comment (alculquicondor) [inline pkg/controller/jobs/raycluster/raycluster_controller_test.go:200]:
```suggestion
		"when workload is admitted and spec.active is set to false, the workload's conditions is set to Evicted": {
```

Or you could remove this test and keep the other.

File: pkg/controller/jobs/raycluster/raycluster_controller_test.go
Comment (alculquicondor) [reply pkg/controller/jobs/raycluster/raycluster_controller_test.go:72]: This is still unresolved.

**Review by alculquicondor — COMMENTED** [review]
/approve
/hold
Leaving LGTM to @astefanutti

File: pkg/controller/jobs/raycluster/raycluster_controller.go
Comment (astefanutti) [inline pkg/controller/jobs/raycluster/raycluster_controller.go:189]: nit: Maybe inline the call to `jobframework.IsOwnerManagedByKueue` directly, and remove that function as it's been generalised.

**Review by astefanutti — APPROVED** [review]
LGTM. Great job, thanks!

This would be good to add the test case covering the injection of node selectors from flavors as pointed out by @alculquicondor. Though I'd rather vote to have this feature in the upcoming release, and have it done as a follow-up, over blocking this PR any longer and miss the release.

**Comment by alculquicondor** [comment]
> Though I'd rather vote to have this feature in the upcoming release

Did you mean the feature or the test?
Node selectors are fundamental to the way flavors work, so we shouldn't miss the feature. I would be cautious about leaving the unit test for later.

**Comment by alculquicondor** [comment]
So it doesn't currently work? I'm rather opposed to merging without it.

/hold

File: test/integration/controller/jobs/raycluster/raycluster_controller_test.go
Comment (alculquicondor) [inline test/integration/controller/jobs/raycluster/raycluster_controller_test.go:200]: ok, it is tested here. It would be useful to have this in unit tests, but it's ok to leave for a follow up

cc @astefanutti 

/hold cancel.

File: pkg/controller/jobs/raycluster/raycluster_controller_test.go
Comment (astefanutti) [inline pkg/controller/jobs/raycluster/raycluster_controller_test.go:74]: Should this should be removed from the test input, as it's expected to be added during the reconciliation?

File: pkg/controller/jobs/raycluster/raycluster_controller_test.go
Comment (alculquicondor) [inline pkg/controller/jobs/raycluster/raycluster_controller_test.go:84]: btw, we have `utiltesting.MakePodSet`

File: pkg/controller/jobs/raycluster/raycluster_controller_test.go
Comment (alculquicondor) [reply pkg/controller/jobs/raycluster/raycluster_controller_test.go:74]: Also a flavor needs to be defined, because that's where the selectors are obtained from.

File: pkg/controller/jobs/raycluster/raycluster_controller_test.go
Comment (alculquicondor) [inline pkg/controller/jobs/raycluster/raycluster_controller_test.go:134]: Alternatively, the admission check could provide some selectors to inject.

**Review by astefanutti — APPROVED** [review]
LGTM

**Comment by alculquicondor** [comment]
/lgtm

based on https://github.com/kubernetes-sigs/kueue/pull/1520#pullrequestreview-1845338497

**Comment by alculquicondor** [comment]
The periodic E2E tests are healthy... so could it be that this PR is breaking them?

/test pull-kueue-test-e2e-main-1-26
/test pull-kueue-test-e2e-main-1-27
/test pull-kueue-test-e2e-main-1-28 
/test pull-kueue-test-e2e-main-1-29

**Comment by alculquicondor** [comment]
tests are definitely flaky https://github.com/kubernetes-sigs/kueue/issues/1658

**Pull request merged**
